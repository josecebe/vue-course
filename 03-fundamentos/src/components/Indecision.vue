<template>
    <img v-if="imgUrl" :src="imgUrl" alt="bg" />
    <div class="bg-dark"></div>

    <div class="indecision-container">
        <input data-test-id="input" type="text" placeholder="Make a question" v-model="question" />
        <p>Remember to end your question using a interrogation symbol '?'</p>

        <div>
            <h2>{{question}}</h2>
            <h1 v-if="isValidQuestion">{{answer}}</h1>
        </div>
    </div>
</template>

<script>
export default {
    name: 'Indecision',

    data() {
        return {
            question: null,
            answer: null,
            imgUrl: null,
            isValidQuestion: false,
        }
    },

    methods: {
        async getAnswer() {
            try {
                this.answer = 'Thinking...'

                const {answer, image} = await fetch('https://yesno.wtf/api')
                    .then(r => r.json())

                switch(answer) {
                    case 'yes':
                        this.answer = '¡Sí!';
                        break;
                    case 'no':
                        this.answer = '¡No!';
                        break;
                    case 'maybe':
                        this.answer = '¡Puede ser!';
                        break;
                }

                this.imgUrl = image;
                this.isValidQuestion = true;

            } catch(error) {
                console.log('IndecisionComponent: API is down');
                this.answer = 'Couldn\'t be loaded';
                this.imgUrl = null;
            }
        },
    },

    watch: {
        question(newValue, oldValue) {
            this.isValidQuestion = false;
            console.log({newValue});

            if (!newValue.endsWith('?')) return;

            this.getAnswer();
        }
    }
}
</script>

<style scoped>
    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }
</style>
