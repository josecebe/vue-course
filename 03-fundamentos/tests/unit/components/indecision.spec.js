import {shallowMount} from "@vue/test-utils";
import Indecision from "@/components/Indecision.vue";

describe('Indecision Tests', () => {
    let wrapper;
    let clgSpy;
    let getAnswerSpy;

    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }));

    beforeEach(() => {
        wrapper = shallowMount(Indecision);
        clgSpy = jest.spyOn(console, 'log');
        getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');
        jest.clearAllMocks();
    });

    test('Should match the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('Writing on the input should not fire the fetch request (console.log)', async() => {
        const input = wrapper.find('[data-test-id="input"]');

        await input.setValue('Hello World');

        expect(clgSpy).toHaveBeenCalledTimes(1);
        expect(getAnswerSpy).not.toHaveBeenCalled();
    });

    test('Writing "?" on the input should fire getAnswer method', async() => {
        const input = wrapper.find('[data-test-id="input"]');

        await input.setValue('Hellow World?');

        expect(getAnswerSpy).toHaveBeenCalled();
    });

    test('getAnswer method test - API works', async() => {
        await wrapper.vm.getAnswer();

        const img = wrapper.find('img');

        expect(img.exists()).toBeTruthy();
        expect(wrapper.vm.imgUrl).toBe('https://yesno.wtf/assets/yes/2.gif');
        expect(wrapper.vm.answer).toBe('¡Sí!');
    });

    test('getAnswer method test - API fails', async() => {
        fetch.mockImplementationOnce(() => Promise.reject('IndecisionComponent: API is down'));
        await wrapper.vm.getAnswer();

        const img = wrapper.find('img');

        expect(img.exists()).toBeFalsy();
        expect(wrapper.vm.answer).toBe('Couldn\'t be loaded');
    });
});
