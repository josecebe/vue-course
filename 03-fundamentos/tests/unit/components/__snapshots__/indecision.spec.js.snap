// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Tests Should match the snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input data-test-id="input" type="text" placeholder="Make a question">
  <p>Remember to end your question using a interrogation symbol '?'</p>
  <div>
    <h2></h2>
    <!--v-if-->
  </div>
</div>
`;
